### PR TITLE
Update worker docs with recent enhancements 

### PR DESCRIPTION
**Pull request by erikjohnston**
We also add some of the new config options to the sample config.

This also updates the wording to recommend using Redis and removing suggestion that it is experimental.

Apologies in advance that me bad at words.

### CODE REVIEW

File: docs/workers.md [hunk 1]
```diff
@@ -242,40 +186,141 @@ endpoints matching the following regular expressions:
...
+load balancing can be done in different ways.
+
+For `/sync` and `/initialSync` requests it will be more efficient if all
+requests from a particular user are routed to a single instance. Extracting a
+user ID from the access token or `Authorization` header is currently left as an
+exercise for the reader. Admins may additionally wish to separate out `/sync`
+requests that have a `since` query parameter from those that don't (and
+`/initialSync`), as requests that don't are known as "initial sync" that happens
+when a user logs in on a new device and can be *very* resource intensive, so
+isolating these requests will stop them from intefering with other users ongoing
```
Comment (babolivier) [inline docs/workers.md:252]:
```suggestion
isolating these requests will stop them from interfering with other users ongoing
```

File: docs/workers.md [hunk 2]
```diff
@@ -64,21 +72,21 @@ configuration; etc.  You should minimise the number of overrides though to
...
+replication port.
 
 For example:
 
 ```yaml
-worker_app: synapse.app.synchrotron
+worker_app: synapse.app.generic_worker
+worker_name: worker1
 
 # The replication listener on the synapse to talk to.
```
Comment (clokep) [inline docs/workers.md:88]:
```suggestion
# The replication listener on the master synapse process.
```

File: docs/workers.md [hunk 3]
```diff
@@ -104,49 +112,6 @@ recommend the use of `systemd` where available: for information on setting up
...
-   according to the `server_name` of your synapse server.)
-2. Check Redis is running and accessible: you should be able to `echo PING | nc -q1
-   localhost 6379` and get a response of `+PONG`.
-3. Install the python prerequisites. If you installed synapse into a
-   virtualenv, this can be done with:
-   ```sh
-   pip install matrix-synapse[redis]
-   ```
-   The debian packages from matrix.org already include the required
-   dependencies.
```
Comment (clokep) [inline docs/workers.md:135]: I suspect that some of this is still useful (e.g. `pip install matrix-synapse[redis]` at least)!

File: docs/workers.md (hunk 1 above)
Comment (clokep) [inline docs/workers.md]:
```suggestion
exercise for the reader.

Admins may additionally wish to separate out `/sync` requests without a `since`
query parameter (and `/initialSync`) -- these requests are the "initial sync" that
happens when a user logs in on a new device and can be *very* resource intensive.
Isolating these requests will stop them from interfering with other users' incremental
```

File: docs/workers.md [hunk 4]
```diff
@@ -242,40 +186,141 @@ endpoints matching the following regular expressions:
...
+should be balanced by source IP so that transactions from the same remote server
+go to the same process.
 
-The `^/_matrix/federation/v1/send/` endpoint must only be handled by a single
-instance.
+Registration/login requests can be handled separately purely to help ensure that
+unexpected load doesn't effect new logins and sign ups.
 
-Note that `federation` must be added to the listener resources in the worker config:
+Finally, event sending requests can be  balanced by the embedded room ID (or
```
Comment (clokep) [inline docs/workers.md:264]:
```suggestion
Finally, event sending requests can be balanced by the embedded room ID (or
```

File: docs/workers.md [hunk 5]
```diff
@@ -242,40 +186,141 @@ endpoints matching the following regular expressions:
...
+streams_writers:
+    events: event_persister1
 ```
 
+
+### `synapse.app.pusher`
+
+Handles sending push notifications to sygnal and email. Doesn't handle any
+REST endpoints itself, but you should set `start_pushers: False` in the
+shared configuration file to stop the main synapse sending these notifications.
```
Comment (clokep) [inline docs/workers.md:296]:
```suggestion
shared configuration file to stop the main synapse sending push notifications.
```

File: docs/workers.md [hunk 6]
```diff
@@ -242,40 +186,141 @@ endpoints matching the following regular expressions:
...
+
+Note this worker cannot be load-balanced: only one instance should be active.
+
+### `synapse.app.appservice`
+
+Handles sending output traffic to Application Services. Doesn't handle any
+REST endpoints itself, but you should set `notify_appservices: False` in the
+shared configuration file to stop the main synapse sending these notifications.
```
Comment (clokep) [inline docs/workers.md:304]:
```suggestion
shared configuration file to stop the main synapse sending appservice notifications.
```

File: docs/workers.md [hunk 7]
```diff
@@ -242,40 +186,141 @@ endpoints matching the following regular expressions:
...
 ### `synapse.app.federation_sender`
 
 Handles sending federation traffic to other servers. Doesn't handle any
 REST endpoints itself, but you should set `send_federation: False` in the
 shared configuration file to stop the main synapse sending this traffic.
 
-Note this worker cannot be load-balanced: only one instance should be active.
+Note that if running multiple federation senders then you must list each
+instance in the `federation_sender_instances` option by their `worker_name`. If
+you add or remove instances they must all be stopped and started together. For example:
```
Comment (clokep) [inline docs/workers.md:317]:
```suggestion
If running multiple federation senders then you must list each
instance in the `federation_sender_instances` option by their `worker_name`.
All instances must be stopped and started when adding or removing instances.
For example:
```

File: synapse/config/redis.py [hunk 8]
```diff
@@ -21,7 +21,7 @@ class RedisConfig(Config):
     section = "redis"
 
     def read_config(self, config, **kwargs):
-        redis_config = config.get("redis", {})
+        redis_config = config.get("redis", {}) or {}
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/config/redis.py:24]: Nope

File: docs/workers.md [hunk 9]
```diff
@@ -18,14 +18,22 @@ running PostgreSQL.
...
-workers so they can be kept in sync with the master process and database state.
+The processes communicate with each other via a Synapse-specific protocol called
+'replication' (analogous to MySQL- or Postgres-style database replication) which
+feeds streams of newly written data between processes so they can be kept in
+sync with the database state.
+
+Additionally, processes may make HTTP requests to each other. Typically this is
+used for operations which need to wait for a reply - such as sending an event.
+
+As of Synapse v1.13.0, it is possible to configure Synapse to send replication
```
Comment (clokep) [inline docs/workers.md:29]: It could be nice to have a snazzy ASCII art diagram here of the processes communicating...

File: synapse/config/workers.py [hunk 10]
```diff
@@ -105,6 +105,27 @@ def read_config(self, config, **kwargs):
                     % (instance, stream)
                 )
 
+    def generate_config_section(self, config_dir_path, server_name, **kwargs):
+        return """\
+        ## Workers ##
+
+        # When using workers this should be a map from `worker_name` to the
+        # HTTP replication listener of the worker, if configured.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/config/workers.py:113]: Yup, exact;y

File: synapse/config/workers.py [hunk 11]
```diff
@@ -105,6 +105,27 @@ def read_config(self, config, **kwargs):
...
+        #
+        #instance_map:
+        #  worker1:
+        #    host: localhost
+        #    port: 8034
+
+        # Experimental: When using workers you can define if workers should
+        # handle event persistence or typing notifications here. Any worker
+        # specified here must also be in the `instance_map`.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/config/workers.py:122]: Yes, though in future these will also be shardable. I guess its a bit inconsistent...

File: synapse/config/workers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/config/workers.py:113]: 👍  I think this is much clearer from reading the workers.md doc. Looks like we link to that already though!

File: docs/workers.md [hunk 12]
```diff
@@ -18,14 +18,22 @@ running PostgreSQL.
...
+via a [Redis pub/sub channel](https://redis.io/topics/pubsub), and is now the
+recommended way of configuring replication. This is an alternative to the old
+direct TCP connections to the master: rather than all the workers connecting to
+the master, all the workers and the master connect to Redis, which relays
+replication commands between processes. This can give a significant cpu saving
+on the master and will be a prerequisite for upcoming performance improvements.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply docs/workers.md:35]: Possibly. I think we should but equally I'm tempted to punt that to a separate discussion of if/when/how we deprecate it.

File: docs/workers.md (hunk 3 above)
Comment (erikjohnston) [reply docs/workers.md:135]: Added back in the section talking about the python dependencies.

File: synapse/config/federation.py [hunk 13]
```diff
@@ -95,4 +95,13 @@ def generate_config_section(self, config_dir_path, server_name, **kwargs):
...
           - 'fc00::/7'
+
+        # If running with federation sender worker instances then they should
+        # be listed by their `worker_name` here, as well as setting
+        # `send_federation` to false.
+        #
+        #send_federation: false
+        #
+        #federation_sender_instances:
+        #  - federation_sender1
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/config/federation.py]: How about now? I'm not 100% convinced about the new wording

File: docs/workers.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (clokep) [reply docs/workers.md]: That's fair enough.

File: synapse/config/federation.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/config/federation.py]: I think it is much clearer. Reading workers.md also helps a lot. :)

**Comment by erikjohnston** [comment]
This needs repointing at release branch.

File: docs/sample_config.yaml [hunk 14]
```diff
@@ -614,11 +614,15 @@ federation_ip_range_blacklist:
   - 'fe80::/64'
   - 'fc00::/7'
 
-# If running with federation sender worker instances then they should
-# be listed by their `worker_name` here, as well as setting
-# `send_federation` to false.
+# Uncomment if using a federation sender worker.
```
Comment (richvdh) [inline docs/sample_config.yaml:617]: could you document what it actually does?

File: docs/sample_config.yaml [hunk 15]
```diff
@@ -614,11 +614,15 @@ federation_ip_range_blacklist:
...
 #
 #send_federation: false
+
+# Multiple federation sender workers can be run, in which case the work is sharded
+# between them. Note that this config must be shared between all instances, and if
+# changed all federation sender workers must be stopped at the same time and then
+# started, to ensure that all instances are running with the same config (otherwise
+# events may be dropped).
```
Comment (richvdh) [inline docs/sample_config.yaml:625]: Prefer the active voice:

```suggestion
# It is possible to run multiple federation sender workers, in which case the work is 
# balanced across them.
#
# This configuration must be shared between all federation sender workers, and if
# changed all federation sender workers must be stopped at the same time and then
# started, to ensure that all instances are running with the same config (otherwise
# events may be dropped).
```

File: docs/sample_config.yaml [hunk 16]
```diff
@@ -614,6 +614,19 @@ federation_ip_range_blacklist:
...
+#
+#send_federation: false
+
+# Multiple federation sender workers can be run, in which case the work is sharded
+# between them. Note that this config must be shared between all instances, and if
+# changed all federation sender workers must be stopped at the same time and then
+# started, to ensure that all instances are running with the same config (otherwise
+# events may be dropped).
+#
+#federation_sender_instances:
```
Comment (richvdh) [inline docs/sample_config.yaml:627]: shouldn't this be in the "Workers" section?

File: docs/sample_config.yaml [hunk 17]
```diff
@@ -2402,3 +2415,40 @@ opentracing:
...
+# HTTP replication listener of the worker, if configured.
+#
+#instance_map:
+#  worker1:
+#    host: localhost
+#    port: 8034
+
+# Experimental: When using workers you can define which worker should
+# handle event persistence or typing notifications. Any worker
+# specified here must also be in the `instance_map`.
```
Comment (richvdh) [inline docs/sample_config.yaml:2432]:
```suggestion
# Experimental: When using workers you can define which workers should
# handle event persistence and typing notifications. Any worker
# specified here must also be in the `instance_map`.
```

File: docs/sample_config.yaml [hunk 18]
```diff
@@ -2402,3 +2415,40 @@ opentracing:
...
+#
+#stream_writers:
+#  events: worker1
+#  typing: worker1
+
+
+# Configuration for Redis when using workers.
+#
+redis:
+  # Uncomment the below to enable using Redis to replicate data between workers.
```
Comment (richvdh) [inline docs/sample_config.yaml:2442]: what happens if it's not enabled?

File: docs/workers.md (hunk 9 above)
Comment (richvdh) [reply docs/workers.md:29]: +1

File: docs/workers.md [hunk 19]
```diff
@@ -16,16 +16,36 @@ workers only work with PostgreSQL-based Synapse deployments. SQLite should only
...
+
+If using Redis the appropriate dependencies must be installed. If using a
+virtual env these can be installed by:
+
+```sh
+pip install matrix-synapse[redis]
+```
+
+Note that these dependencies are included if synapse was installed with `[all]`,
+or if using the debian packages from matrix.org.
```
Comment (richvdh) [inline docs/workers.md:46]: this probably belongs in the "configuration" section (which you might consider renaming to "Setting up workers" or something) rather than here which feels more like an overview of how it works).

File: docs/workers.md [hunk 20]
```diff
@@ -16,16 +16,36 @@ workers only work with PostgreSQL-based Synapse deployments. SQLite should only
...
+Redis, which relays replication commands between processes. This can give a
+significant cpu saving on the main process and will be a prerequisite for
+upcoming performance improvements.
+
+If using Redis the appropriate dependencies must be installed. If using a
+virtual env these can be installed by:
+
+```sh
+pip install matrix-synapse[redis]
+```
```
Comment (richvdh) [inline docs/workers.md:43]: I realise it might be considered out of scope, but I think a few words about installing redis might not go amiss.  basically the old bullet points 1 and 2.

File: docs/workers.md [hunk 21]
```diff
@@ -35,26 +55,26 @@ the correct worker, or to the main synapse instance. Note that this includes
...
 ```
 
-Under **no circumstances** should these replication API listeners be exposed to
-the public internet; they have no authentication and are unencrypted.
+See the sample config for the full documentation of each option.
+
+Under **no circumstances** should the replication API listener be exposed to the
+public internet; it has no authentication and is unencrypted.
 
 You should then create a set of configs for the various worker processes.  Each
```
Comment (richvdh) [inline docs/workers.md:79]:
```suggestion
You should then create a configuration file for each worker process.  Each
```

File: docs/workers.md [hunk 22]
```diff
@@ -35,26 +55,26 @@ the correct worker, or to the main synapse instance. Note that this includes
 requests made to the federation port. See [reverse_proxy.md](reverse_proxy.md)
 for information on setting up a reverse proxy.
 
-To enable workers, you need to add *two* replication listeners to the
-main Synapse configuration file (`homeserver.yaml`). For example:
+To enable workers, you need to add both a HTTP replication listener and redis
+config to the main Synapse configuration file (`homeserver.yaml`). For example:
```
Comment (richvdh) [inline docs/workers.md:59]: should we describe `homeserver.yaml` as a "shared configuration file" or something?

Comment (erikjohnston) [reply docs/workers.md:59]: Good point, now that we're using "main" to mean the main synapse process.

File: docs/sample_config.yaml (hunk 18 above)
Comment (erikjohnston) [reply docs/sample_config.yaml:2445]: Hmm, I guess we should reword this to make it clear that it needs (unless using old style replication) to be enabled if using workers

File: docs/workers.md (hunk 20 above)
Comment (erikjohnston) [reply docs/workers.md:52]: I've add some words back in

File: docs/workers.md [hunk 23]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
+
+The inbound federation transaction request `^/_matrix/federation/v1/send/`
+should be balanced by source IP so that transactions from the same remote server
+go to the same process.
+
+Registration/login requests can be handled separately purely to help ensure that
+unexpected load doesn't effect new logins and sign ups.
+
+Finally, event sending requests can be balanced by the embedded room ID (or
+URI, or even just round robin). If there is a large bridge connected that is
```
Comment (richvdh) [inline docs/workers.md:283]: what do you mean by "embedded room ID"?

File: docs/workers.md [hunk 24]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
 
-Note that `federation` must be added to the listener resources in the worker config:
+Federation and client requests can be balanced via simple round robin.
+
+The inbound federation transaction request `^/_matrix/federation/v1/send/`
+should be balanced by source IP so that transactions from the same remote server
+go to the same process.
+
+Registration/login requests can be handled separately purely to help ensure that
+unexpected load doesn't effect new logins and sign ups.
```
Comment (richvdh) [inline docs/workers.md:280]: I feel I might have mentioned this one in the past.

```suggestion
unexpected load doesn't affect new logins and sign ups.
```

Comment (richvdh) [reply docs/workers.md:280]: do we actually do this on m.org? I thought they just went to the client reader pool?

File: docs/workers.md [hunk 25]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
+
+Finally, event sending requests can be balanced by the embedded room ID (or
+URI, or even just round robin). If there is a large bridge connected that is
+sending or may send lots of events, then a dedicated set of workers can be
+provisioned to ensure that bursts or increases of event sending is isolated from
+effecting events sent by real users.
```
Comment (richvdh) [inline docs/workers.md:286]:
```suggestion
provisioned to limit the effects of bursts of events from that bridge on events sent by 
normal users.
```

File: docs/workers.md [hunk 26]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
+
+#### Stream writers
+
+Additionally, there is *experimental* support for moving writing of specific
+streams (such as events and typing) off of the main process to a particular
+worker. This requires use of Redis.
```
Comment (richvdh) [inline docs/workers.md:292]:
```suggestion
worker. (This is only supported with Redis-based replication.)
```

File: docs/workers.md [hunk 27]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
+
+To enable this then the worker must have a HTTP replication listener configured,
```
Comment (richvdh) [inline docs/workers.md:294]:
```suggestion
To enable this, the worker must have a HTTP replication listener configured,
```

File: docs/workers.md [hunk 28]
```diff
@@ -242,40 +204,142 @@ endpoints matching the following regular expressions:
...
-   resources:
-     - names:
-       - federation
+instance_map:
+    event_persister1:
+        host: localhost
+        port: 8034
+
+streams_writers:
+    events: event_persister1
```
Comment (richvdh) [inline docs/workers.md:306]: which streams can be moved in this way? it says "such as events and typing" above but I'm not sure if that is complete?

File: docs/workers.md (hunk 9 above)
Comment (richvdh) [reply docs/workers.md:29]: does https://matrix.sw1v.org/_matrix/media/r0/download/sw1v.org/uhqIFIqONRWcRsKgVvbehsqF help? some of the links have been left out for clarity but I think it might help?

Comment (erikjohnston) [reply docs/workers.md:29]: Possibly? I'm not really sure who it's aimed at.

Comment (erikjohnston) [reply docs/workers.md:29]: Am happy to put it at the end of the doc, to be clear.

File: docs/workers.md [hunk 29]
```diff
@@ -104,54 +130,11 @@ recommend the use of `systemd` where available: for information on setting up
...
-    `password`. Normally none of these are required.
-5. Restart master and all workers.
-
-Once redis replication is in use, `worker_replication_port` is redundant and
-can be removed from the worker configuration files. Similarly, the
-configuration for the `listener` for the TCP replication port can be removed
-from the main configuration file. Note that the HTTP replication port is
-still required.
 
 ### Using synctl
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply docs/workers.md]: I've moved this to a separate doc.

File: docs/workers.md (hunk 9 above)
Comment (richvdh) [reply docs/workers.md:29]: > I'm not really sure who it's aimed at.

well, people trying to set up workers. The idea is that a picture helps people get an overview of how it works.

File: docs/workers.md [hunk 30]
```diff
@@ -277,21 +248,23 @@ should be balanced by source IP so that transactions from the same remote server
...
 #### Stream writers
 
 Additionally, there is *experimental* support for moving writing of specific
-streams (such as events and typing) off of the main process to a particular
-worker. This requires use of Redis.
+streams (such as events) off of the main process to a particular worker. (This
+is only supported with Redis-based replication.)
 
-To enable this then the worker must have a HTTP replication listener configured,
+Currently support streams are `events` and `typing`.
```
Comment (richvdh) [inline docs/workers.md:265]:
```suggestion
Currently supported streams are `events` and `typing`.
```

File: docs/workers.md (hunk 9 above)
Comment (erikjohnston) [reply docs/workers.md:29]: Ok, will add it to the end with a note at the start pointing to it :+1:

**Comment by richvdh** [comment]
hrm, there was more that github has eaten

File: docs/workers.md [hunk 31]
```diff
@@ -16,69 +16,103 @@ workers only work with PostgreSQL-based Synapse deployments. SQLite should only
...
-main Synapse configuration file (`homeserver.yaml`). For example:
+To enable workers you should create a configuration file for each worker
+process. Each worker configuration file inherits the configuration of the shared
+homeserver configuration file.  You can then override configuration specific to
+that worker, e.g. the HTTP listener that it provides (if any); logging
+configuration; etc.  You should minimise the number of overrides though to
+maintain a usable config.
+
+Next you need to add both a HTTP replication listener and redis config to the
+shared Synapse configuration file (`homeserver.yaml`). For example:
```
Comment (richvdh) [inline docs/workers.md:79]:
```suggestion
When using workers, each worker process has its own configuration file which 
contains settings specific to that worker, such as the HTTP listener that it 
provides (if any), logging configuration, etc.

Normally,  the worker processes are configured to read from a shared
configuration file as well as the worker-specific configuration files. This
makes it easier to keep common configuration settings synchronised across all the 
processes.

The main process is somewhat special in this respect: it does not normally
need its own configuration file and can take all of its configuration from the
shared configuration file.

Normally, only a couple of changes are needed to make an existing configuration
file suitable for use with workers. First, you need to enable an "HTTP replication
listener" for the main process; and secondly, you need to enable redis-based
replication. For example:
```
